### PR TITLE
[onert] Introduce TrainableTensor

### PR DESCRIPTION
**Pull request by ragmani**
This commit introduce TrainableTensor into train backend.

ONE-DCO-1.0-Signed-off-by: ragmani <ragmani0216@gmail.com>

### CODE REVIEW

**Comment by ragmani** [comment]
draft #10984

**Comment by glistening** [comment]
@nnfw-bot test onert-x64-release

**Review by jyoungyun — APPROVED** [review]
LGTM

File: runtime/onert/backend/train/Tensor.h [hunk 1]
```diff
@@ -43,7 +44,23 @@ class Tensor : public basic::Tensor
...
+  TrainableTensor(const ir::OperandInfo &info, const ir::Layout layout) : Tensor{info, layout}
+  {
+    // DO NOTHING
+  }
+
+public:
+  void applyGradient(const IPortableTensor &grad_tensor, double lr) override;
+
+public:
+  void fillBuffer(const std::shared_ptr<ir::Data> &data);
```
Comment (glistening) [inline runtime/onert/backend/train/Tensor.h:62]: @ragmani `fillBuffer` is not a part of interface `ITrainableTensor`. Is `fillBuffer` a part of `TrainableTensor` only?

File: runtime/onert/core/include/backend/train/ITrainableTensor.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __ONERT_BACKEND_TRAIN_ITRAINABLE_TENSOR_H__
+#define __ONERT_BACKEND_TRAIN_ITRAINABLE_TENSOR_H__
```
Comment (glistening) [inline runtime/onert/core/include/backend/train/ITrainableTensor.h:18]:
```suggestion
#ifndef __ONERT_BACKEND_TRAIN_I_TRAINABLE_TENSOR_H__
#define __ONERT_BACKEND_TRAIN_I_TRAINABLE_TENSOR_H__
```

File: runtime/onert/core/include/backend/train/ITrainableTensor.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+   * @param lr          Learning rate
+   */
+  virtual void applyGradient(const IPortableTensor &grad_tensor, double lr) = 0;
+};
+
+} // namespace train
+} // namespace backend
+} // namespace onert
+
+#endif // __ONERT_BACKEND_TRAIN_ITRAINABLE_TENSOR_H__
```
Comment (glistening) [inline runtime/onert/core/include/backend/train/ITrainableTensor.h:51]:
```suggestion
#endif // __ONERT_BACKEND_TRAIN_I_TRAINABLE_TENSOR_H__
```

File: runtime/onert/core/include/backend/train/ITrainableTensor.h (hunk 2 above)
Comment (ragmani) [reply runtime/onert/core/include/backend/train/ITrainableTensor.h:18]: There are many styles in onert.
`__ONERT_BACKEND_TRAIN_I_TRAINABLE_TENSOR_H__`
`__ONERT_BACKEND_TRAIN_ITRAINABLE_TENSOR_H__`
`__ONERT_BACKEND_TRAIN_ITRAINABLETENSOR_H__`

I don't know which one to follow.
If there is a guide, I will follow that guide in the future.

Comment (glistening) [reply runtime/onert/core/include/backend/train/ITrainableTensor.h:18]: I only knew the `_I_` styles from onert.

e.g.
- ITensor.h - #ifndef `__ONERT_BACKEND_OPERAND_I_TENSOR_H__` (? old days)
- ICompiler.h - #ifndef `__ONERT_COMPILER_I_COMPILER_H_` (by @hseok-oh)

But I searched after you comment above, then I've found other styles. : )

- IConfig.h - #ifndef `__ONERT_BACKEND_ICONFIG_H__` (? old days)
- IOperation.h - #ifndef `__ONERT_IR_IOPERATION_H__` by you (@ragmani)

It seems the style is mixed. Then, it is okay to keep your style. : )

File: runtime/onert/backend/train/Tensor.h (hunk 1 above)
Comment (ragmani) [reply runtime/onert/backend/train/Tensor.h:62]: Yes, there will be filled in `train` backend. So, I think `ITrainableTensor` does not need to have `fillBuffer` now.

Comment (glistening) [reply runtime/onert/backend/train/Tensor.h:62]: I wondered the similar method would exists in `Tensor`. But it has only `setBuffer(uint8_t buf)`.

**Review by glistening — APPROVED** [review]

**Pull request merged**